apps/src/studio/collidable.js

```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
+  // If not, determine it.
+  if (this.destGridX === undefined || reachedDestinationGridPosition) {
+
+    var sprite = Studio.sprite[0];
+
+    var spriteX = sprite.x + sprite.width/2;
+    var spriteY = sprite.y + sprite.height/2;
+
+    // let's try scoring each square
+    var candidates = [];
+
+    var bufferDistance = 60;
+
+    for (var candidateX = this.gridX - 1; candidateX <= this.gridX + 1; candidateX++) {
+      for (var candidateY = this.gridY - 1; candidateY <= this.gridY + 1; candidateY++) {
+        candidate = {gridX: candidateX, gridY: candidateY};
+        candidate.score = 0;
+
+        if (type == "roamGrid") {
+          candidate.score ++;
+        } else if (type == "chaseGrid") {
+          if (candidateY == this.gridY - 1 && spriteY < this.y - bufferDistance) {
+            candidate.score ++;
+          } else if (candidateY == this.gridY + 1 && spriteY > this.y + bufferDistance) {
+            candidate.score ++;
+          }
+
+          if (candidateX == this.gridX - 1 && spriteX < this.x - bufferDistance) {
+            candidate.score ++;
+          } else if (candidateX == this.gridX + 1 && spriteX > this.x + bufferDistance) {
+            candidate.score ++;
+          }
+        } else if (type == "fleeGrid") {
+          candidate.score = 1;
+          if (candidateY == this.gridY - 1 && spriteY > this.y - bufferDistance) {
+            candidate.score ++;
+          } else if (candidateY == this.gridY + 1 && spriteY < this.y + bufferDistance) {
+            candidate.score ++;
+          }
+
+          if (candidateX == this.gridX - 1 && spriteX > this.x - bufferDistance) {
+            candidate.score ++;
+          } else if (candidateX == this.gridX + 1 && spriteX < this.x + bufferDistance) {
+            candidate.score ++;
+          }
+        }
+
+        if (candidate.score > 0) {
+          Studio.drawDebugRect(
+            "roamGridPossibleDest", 
+            candidateX * Studio.SQUARE_SIZE + Studio.HALF_SQUARE, 
+            candidateY * Studio.SQUARE_SIZE + Studio.HALF_SQUARE, 
+            Studio.SQUARE_SIZE, 
+            Studio.SQUARE_SIZE);
+        }
+        candidates.push(candidate);
+      }
+    }
+
+    // cull candidates that won't be possible
+    for (var i = candidates.length-1; i >= 0; i--) {
+      var candidate = candidates[i];
+      var atEdge = candidate.gridX < 0 || candidate.gridX >= Studio.COLS ||
+                   candidate.gridY < 0 || candidate.gridY >= Studio.ROWS;
+      var hasWall = !atEdge && Studio.map[candidate.gridY][candidate.gridX] & SquareType.WALL;
+      if (atEdge || hasWall || candidate.score === 0) {
+        candidates.splice(i, 1);
+      }
+    }
+
+    if (candidates.length > 0) {
+      // shuffle everything (so that even scored items are shuffled, even after the sort)
+      candidates = _.shuffle(candidates);
+
+      // then sort everything based on score.
+      candidates.sort(function (a, b) {
+        return b.score - a.score;
+      });
```

Comment:
Not sure how many candidates we expect to have, but it seems like it would be more efficient to sort, pull out the top candidates, and then only shuffle those (since it looks like we just want one of the top candidates). Could be that this is an unnecessary perf optimization.